### PR TITLE
Position tracker

### PR DESCRIPTION
**Pull request by dalejung**
Moved out position tracker.

TODO:
 
- [x] change out filter/map to use six
- [x] create das custom error.
- [x] cache position_values
- [x] remove np.dot and just sum position_values

Maybe:

- [ ] cython the whole bizness?

### CODE REVIEW

**Comment by dalejung** [comment]
i bet this fails flake8...

**Comment by llllllllll** [comment]
you can run that locally ;)

File: zipline/finance/performance/period.py [hunk 1]
```diff
@@ -72,41 +72,41 @@
...
 from six import iteritems, itervalues
 
 import zipline.protocol as zp
-from . position import positiondict
 
 log = logbook.Logger('Performance')
+TRADE_TYPE = zp.DATASOURCE_TYPE.TRADE
 
+def position_proxy(func):
+    def _proxied(self, *args, **kwargs):
```
Comment (llllllllll) [inline zipline/finance/performance/period.py]: `@wraps(f)` so that the method name is not lost.

**Comment by dalejung** [comment]
lol, yeah but it fails through qexec hm. I guess if I'm not passed in a PositionTracker, I should make one to make 100% backwards compat.

**Comment by dalejung** [comment]
Also, the position_proxy and proxy_error were meant as developmental stopgaps. though, if I replace them it'd be too much typing :(

File: zipline/finance/performance/period.py [hunk 2]
```diff
@@ -72,41 +72,41 @@
...
-            self.series = \
-                self.series.append(
-                    pd.Series({key: value}))
-            self._loc_map = dict(
-                zip(self.series.index,
-                    range(len(self.series))))
-            self.values = self.series.values
+    meth_name = inspect.stack()[1][3]
+    msg = "{meth_name} should have been proxied to position_tracker."
+    return Exception(msg.format(meth_name=meth_name))
```
Comment (llllllllll) [inline zipline/finance/performance/period.py]: Could we make this a specific exception, maybe `ProxyError` that does the inspect stuff?

File: zipline/finance/performance/tracker.py [hunk 3]
```diff
@@ -357,17 +362,15 @@ def check_upcoming_dividends(self, midnight_of_date_that_just_ended):
         pay_date_mask = (self.dividend_frame['pay_date'] == next_trading_day)
         dividends_payable = self.dividend_frame[pay_date_mask]
 
+        if len(dividends_earnable):
```
Comment (llllllllll) [inline zipline/finance/performance/tracker.py]: do we care about the length being truthy or the collection being truthy?

Comment (dalejung) [inline zipline/finance/performance/tracker.py]: len, the `pay_date_mask` is already grabbing the truthy.

**Comment by dalejung** [comment]
HM. I'm starting to think that there needs to be something between perf tracker and position_tracker/performance_periods. Either that or perf tracker needs to be more generalized. the qexec bencharkmark source uses PerformancePeriod as a mini perf tracker. 

I don't like how splitting out the position management requires me to coordinate a position_tracker and perf_period. hmsies.

**Comment by dalejung** [comment]
```
-        pp = perf.PerformancePeriod(1000.0)
-        pp.execute_transaction(txn1)
-        pp.execute_transaction(txn2)
+        pt = perf.PositionTracker()
+        pp = perf.PerformancePeriod(1000.0, position_tracker=pt)
+        pt.execute_transaction(txn1)
+        pp.handle_execution(txn1)
+        pt.execute_transaction(txn2)
+        pp.handle_execution(txn2)
```

is the crap i'm talking abouts.

**Comment by dalejung** [comment]
woo, tracking down serialization bugs is fun.

**Comment by dalejung** [comment]
closed in favor of https://github.com/quantopian/zipline/pull/505

**Pull request closed without merging**